Disallow async IBAction methods

etcwilde opened the pull request:
The IBAction signature does not allow completion handlers, so making
IBAction methods async doesn't make sense. Disallowing it statically so
it doesn't just break people when they try.

Resolves: rdar://83259194

etcwilde committed:
Disallow async IBAction methods

The IBAction signature does not allow completion handlers, so making
IBAction methods async doesn't make sense. Disallowing it statically so
it doesn't just break people when they try.

beccadax commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1433,6 +1433,8 @@ ERROR(invalid_ibinspectable,none,
 ERROR(invalid_ibaction_decl,none,
       "only instance methods can be declared @%0", (StringRef))
+ERROR(invalid_ibaction_decl_async,none,
+      "@%0 instance methods cannot be async", (StringRef))
```
Nit: The other diagnostics all just say `@%0 methods`; let's do that here too.

```suggestion
      "@%0 methods cannot be async", (StringRef))
```

SerenaKit replied:
Should probably also rename the diagnostic because the message isn’t hardcoded 🙃

etcwilde replied:
Done

beccadax reviewed (changes requested):
You have the right idea, but I have some nitpicky requests/suggestions to make.

beccadax commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1433,6 +1433,8 @@ ERROR(invalid_ibinspectable,none,
       "only instance properties can be declared @%0", (StringRef))
 ERROR(invalid_ibaction_decl,none,
       "only instance methods can be declared @%0", (StringRef))
+ERROR(invalid_ibaction_decl_async,none,
```
This error only applies to `@IBAction` (and `@IBSegueAction`, which is treated as a variant), so it should mention that in the name. The other diagnostics emitted for this attribute are all of the form `invalid_ibaction_<incorrect trait>` with an optional `_<phrasing variant>` suffix, so the name that would best match the others is probably `invalid_ibaction_async`, but we tend not to be very picky about diagnostic IDs as long as they get the point across.

SerenaKit replied:
I see, will mark this as resolved then

etcwilde replied:
The name changed when I merged this with the same error for `@_cdecl` attributed functions.

beccadax commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -530,6 +530,12 @@ void AttributeChecker::visitIBActionAttr(IBActionAttr *attr) {
+  if (FD->isAsyncContext()) {
+    diagnoseAndRemoveAttr(attr, diag::invalid_ibaction_decl_async,
+                          attr->getAttrName());
+    return;
```
(To be clear, if it's *easy* to wrap the body in something like `Task`—perhaps we already have a helper for that somewhere?—I would think about doing that. If it's gonna be a hassle, just recommend removing `async` and make sure that the diagnostics on using `await` in a synchronous method body won't try to put the `async` back on an `@IBAction` or `@IBSegueAction` method.)

etcwilde replied:
Yeah, I was kind of thinking that as I wrote it. I'll try to figure out how to do that transformation.

etcwilde replied:
Done

etcwilde committed:
Merge invalid async decl + attribute

This patch merges the two diagnostics complaining about asynchronous
function declarations with incompatible attributes. The two are cdecl
and ibaction at the moment.

etcwilde committed:
Add fix recommendation

This adds a note with a possible way to fix the issue, recommending
removing the `async` effect and wrapping the body of the function in a
task.

etcwilde committed:
Emit fix-it for diagnostic

This patch teaches the compiler to emit a fix-it to automatically wrap
everything in the function in a Task and remove the effects from the
function declaration.

etcwilde committed:
Format patches

etcwilde commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -530,6 +530,12 @@ void AttributeChecker::visitIBActionAttr(IBActionAttr *attr) {
     return;
   }
 
+  if (FD->isAsyncContext()) {
```
Moved. This made the fix-it a bit more challenging to write since I actually needed to keep the return type for the segue action.

etcwilde replied:
Done

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

beccadax reviewed (approved):
Perfect! Thank you!

etcwilde commented:
@swift-ci please test and merge
